csrc/id_model/id_model.cpp

```diff
@@ -611,16 +671,187 @@ void IdModel::buildLoopGraph() {
+  //    domain may not be loop mapped as mentioned above. Instead,
+  //    find all loop-mapped domains with the broadcast domain and
+  //    pick one that is exactly mapped with the resolving domain
+  //
+  // Note again this process is only done for root domains. Once we
+  // find promotion relationships for root domains, we propagate the
+  // mappings to derived domains
+  for (const ValGroup& iel_group : iel_graph.disjointValSets().disjointSets()) {
+    NVF_ERROR(!iel_group->empty());
+
+    IterDomain* iel_group_id = iel_group->front()->as<IterDomain>();
+
+    if (!iel_group_id->isBroadcast()) {
+      continue;
+    }
+
+    // Collect all the exact groups of the resolutions of the broadcast id's
+    ValGroups resolved_exact_groups;
+    for (Val* bcast_id : *iel_group) {
+      if (auto p2c_root_broadcast_resolution_map_it =
+              info.p2c_root_broadcast_resolution_map.find(
+                  bcast_id->as<IterDomain>());
+          p2c_root_broadcast_resolution_map_it !=
+          info.p2c_root_broadcast_resolution_map.end()) {
+        resolved_exact_groups.pushBack(
+            idGraph(IdMappingMode::EXACT)
+                .toGroups(p2c_root_broadcast_resolution_map_it->second));
+      }
+    }
+
+    if (resolved_exact_groups.empty()) {
+      // No resolution
+      continue;
+    }
+
+    // resolved_exact_groups is a list of IDs that resolves the
+    // broadcast. We only care those that are also in the same loop
+    // group, and there must be just one or none. Otherwise, the
+    // resolution is ambiguous.
+
+    // Collect all the exact groups in the loop set containing this iel_group
+    const ValGroup& loop_group =
+        idGraph(IdMappingMode::LOOP).toGroup(iel_group_id);
+    ValGroups loop_covered_exact_groups =
+        idGraph(IdMappingMode::EXACT).toGroups(*loop_group);
+
+    // The intersection of the exact groups that the broadcast domains can be
+    // broadcasted to, and those that exist within the same loop groop are is
+    // the promotion needed for this iel_group. The promotion should
+    // be none or unique.
+    ValGroups loop_exact_resolved_intersection =
+        resolved_exact_groups.computeIntersect(loop_covered_exact_groups);
+
+    if (loop_exact_resolved_intersection.empty()) {
+      // No promotion
+      continue;
+    }
+
+    if (loop_exact_resolved_intersection.size() > 1) {
+      // Ambiguous promotion. This should not happen.
+      std::stringstream err_msg;
+      err_msg
+          << "Invalid multiple broadcast resolution within shared loops detected, group:\n  "
+          << iel_group->toString() << "\nIs being broadcasted to:";
+      for (const ValGroup& entry : loop_exact_resolved_intersection) {
+        err_msg << "\n  " << entry->toString();
+      }
+      NVF_ERROR(false, err_msg.str());
+    }
+
+    const ValGroup& exact_resolution_group =
+        loop_exact_resolved_intersection.front();
```

Comment:
Should this be just:

```C++
if (resolved_exact_groups.size() > 1) {
  NVF_ERROR(false, err_msg.str());
}
const ValGroup& exact_resolution_group = resolved_exact_groups.front();
```
?

Reply:
Not necessarily, I believe. `resolved_exact_groups` may contain non-broadcast domains where the broadcast domain is not inlined. The loop graph is used to exclude non-inlined cases.

When `loop_exact_resolved_intersection.size()` is 1, `resolved_exact_groups.size()` should also be 1. However, `loop_exact_resolved_intersection.size()` is 0, `resolved_exact_groups.size()` can be greater than 1. For example:

```
auto t0 = [b0];
fusion.addInput(t0);
auto t1 = [i1];
fusoin.addInput(t1);
auto t2 = [i2];
fusion.addInput(t2);
// i1 != i2

auto t3 = t0 + t1
fusion.addOutput(t3);

auto t4 = t0 + t2
fusion.addOutput(t4);
```

This is a valid program. The `b0` broadcast domain is a non-unique broadcast domain. `resolved_exact_groups` would have two groups: `{i1}` and `{i2}`, but  `loop_exact_resolved_intersection` should be empty.